WcaOnRails/app/controllers/results_submission_controller.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+
+class ResultsSubmissionController < ApplicationController
+  before_action :authenticate_user!
+  before_action -> { redirect_to_root_unless_user(:can_submit_competition_results?, competition_from_params) }
+
+  def edit
+    @competition = competition_from_params
+  end
+
+  def submit
+    @competition = competition_from_params
+
+    if params[:results].blank? || params[:message].blank?
+      flash.now[:danger] = "Please make sure to fill in the message and attach the results file."
+      render :edit
+    else
+      results_file = params[:results]
+      results_file_path = "uploads/#{@competition.id}/#{results_file.original_filename}"
+
+      FileUtils.mkdir_p(File.dirname(results_file_path))
+      File.open(results_file_path, "w") { |file| file.write(results_file.read) }
+
+      CompetitionsMailer.results_submitted(@competition, params[:message], results_file_path).deliver_later
```

Comment:
Yes, I wanted to talk about that.
I didn't do it because it doesn't scale and I don't know where the limit is. But I _am_ writing files to disk and I am not deleting them when the email is sent.
I can try to figure out how long a string I can pass here. If there is a chance it's ever going to overflow I'll see if I can delete the file after sending the email.